cpp/include/cuspatial/detail/join/quadtree_point_in_polygon.cuh

```diff
@@ -148,64 +154,84 @@ std::pair<rmm::device_uvector<IndexType>, rmm::device_uvector<IndexType>> quadtr
+  // enumerate pairs of (poly_index, point_index) for each point in each quadrant.
+  //
+  // In Python pseudocode:
+  // ```
+  // pp_pairs = []
+  // for polygon, quadrant in pq_pairs:
+  //   for point in quadrant:
+  //     pp_pairs.append((polygon, point))
+  // ```
+  //
+  auto global_to_poly_and_point_indices = [&](auto offset = 0) {
+    return detail::make_counting_transform_iterator(
+      offset,
       detail::compute_poly_and_point_indices{quad_offsets_iter,
                                              local_point_offsets.begin(),
                                              local_point_offsets.end(),
                                              poly_indices_first});
-
-    auto poly_and_point_indices =
-      thrust::make_zip_iterator(poly_indices.begin(), point_indices.begin()) + num_intersections;
-
-    // Compute the number of intersections by removing (poly, point) pairs that don't intersect
-    num_intersections += thrust::distance(
-      poly_and_point_indices,
-      thrust::copy_if(rmm::exec_policy(stream),
-                      global_to_poly_and_point_indices,
-                      global_to_poly_and_point_indices + chunk_size,
-                      poly_and_point_indices,
-                      detail::test_poly_point_intersection{point_xys_iter, polygons}));
-
-    chunk_offset += chunk_size;
+  };
+
+  auto run_quadtree_point_in_polygon = [&](auto output_size) {
+    // Allocate the output polygon and point index pair vectors
+    rmm::device_uvector<IndexType> poly_indices(output_size, stream);
+    rmm::device_uvector<IndexType> point_indices(output_size, stream);
+
+    auto num_intersections = count_in_chunks([&](auto memo, auto offset, auto size) {
+      auto poly_and_point_indices =
+        thrust::make_zip_iterator(poly_indices.begin(), point_indices.begin()) + memo;
+      // Remove (poly, point) pairs that don't intersect
+      return thrust::distance(poly_and_point_indices,
+                              thrust::copy_if(rmm::exec_policy(stream),
+                                              global_to_poly_and_point_indices(offset),
+                                              global_to_poly_and_point_indices(offset) + size,
+                                              poly_and_point_indices,
+                                              test_poly_point_pair));
+    });
+
+    if (num_intersections < output_size) {
+      poly_indices.resize(num_intersections, stream);
+      point_indices.resize(num_intersections, stream);
+      poly_indices.shrink_to_fit(stream);
+      point_indices.shrink_to_fit(stream);
+    }
+
+    return std::pair{std::move(poly_indices), std::move(point_indices)};
+  };
+
+  try {
+    // First attempt to run the hit test assuming allocating space for all possible intersections
+    // fits into the available memory.
+    return run_quadtree_point_in_polygon(num_total_points);
+  } catch (std::exception const&) {
```

Comment:
You should probably catch the more specific `rmm::out_of_memory` and then throw if any other exception is raised?